File: C#/16. 3Sum Closest.cs
// TC: O(n^2)
// SC: O(1)
public class Solution {
    public int ThreeSumClosest(int[] nums, int target) {
        int len = nums.Length;
        Array.Sort(nums);
        // diff = sum - target
        int diff = Int32.MaxValue;
        for (int i = 0; i < len -2; i++)
        {
            int j = i + 1;
            int k = len - 1;
            while (j < k)
            {
                int sum = nums[i] + nums[j] + nums[k];
                if (Math.Abs(sum - target) < Math.Abs(diff))
                    diff = sum - target;
                if (sum < target)
                    j++;
                else
                    k--;
            }
            if (diff == 0)
                break;
        }
        return diff + target;
    }
}

File: java/16. 3Sum Closest.java
//16. 3Sum Closest
// TC: O(N^2) where N = nums.length
// SC: O(1)

class Solution {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        
        int low = 0, high = nums.length - 1;
        while (low + 1 < high - 1 && nums[low + 1] + nums[high] + nums[high - 1] < target) { // prune left side
            low++;
        }
        while (high - 1 > low + 1 && nums[low] + nums[low + 1] + nums[high - 1] > target) { // prune right side
            high--;
        }
        
        int l = low;
        int closest = nums[low] + nums[low + 1] + nums[low + 2];
        
        while (l < high - 1) {
            int left = l + 1;
            int right = high;
            
            while (left < right) { 
                int sum = nums[l] + nums[left] + nums[right];
                
                if (sum == target) return target;
                
                if (Math.abs(sum - target) < Math.abs(closest - target))
                    closest = sum;

                if (sum < target) {
                    left++;
                } else {
                    right--;
                }
            }
            l++;
        }
        return closest;
    }
}



